File: p341m/nested_iterator.js
// PART: ASSUMPTION
/**
 * // This is the interface that allows for creating nested lists.
 * // You should not implement it, or speculate about its implementation
 */
const Nested = require('nested_integer');

// PART: SOLUTION
/**
 * @constructor
 * @param {NestedInteger[]} nestedList
 */
var NestedIterator = function (nestedList) {
  this.stack = [];
  this.prepareStack(nestedList);
};
NestedIterator.prototype.prepareStack = function (nestedList) {
  for (let i = nestedList.length - 1; i >= 0; --i) {
    this.stack.push(nestedList[i]);
  }
};
/**
 * @this NestedIterator
 * @returns {boolean}
 */
NestedIterator.prototype.hasNext = function () {
  while (
    this.stack.length > 0 &&
    !this.stack[this.stack.length - 1].isInteger()
  ) {
    let l = this.stack.pop().getList();
    this.prepareStack(l);
  }
  return this.stack.length > 0;
};
/**
 * @this NestedIterator
 * @returns {integer}
 */
NestedIterator.prototype.next = function () {
  return this.stack.pop().getInteger();
};
// PART: TEST
/**
 * Your NestedIterator will be called like this:
 * var i = new NestedIterator(nestedList), a = [];
 * while (i.hasNext()) a.push(i.next());
 */
[
  [[[]], []],
  [[1], [1]],
  [[[1, 2]], [1, 2]],
  [[[[3]]], [3]],
  [
    [[1, 1], 2, [1, 1]],
    [1, 1, 2, 1, 1],
  ],
  [
    [1, [4, [[5], 6]]],
    [1, 4, 5, 6],
  ],
  [
    [[[3], 4], [5], 6],
    [3, 4, 5, 6],
  ],
].forEach(([t, expected]) => {
  let nestedList = Array.from(t, (v) => new Nested.NestedInteger(v));
  let [i, actual] = [new NestedIterator(nestedList), []];
  while (i.hasNext()) {
    actual.push(i.next());
  }
  console.log('Nested Iteration of', t, '->', actual);
  console.assert(actual.toString() === expected.toString());
});

File: p341m/nested_iterator.py
# """
# This is the interface that allows for creating nested lists.
# You should not implement it, or speculate about its implementation
# """
# class NestedInteger:
#    def isInteger(self) -> bool:
#        """
#        @return True if this NestedInteger holds a single integer, rather than a nested list.
#        """
#
#    def getInteger(self) -> int:
#        """
#        @return the single integer that this NestedInteger holds, if it holds a single integer
#        Return None if this NestedInteger holds a nested list
#        """
#
#    def getList(self) -> [NestedInteger]:
#        """
#        @return the nested list that this NestedInteger holds, if it holds a nested list
#        Return None if this NestedInteger holds a single integer
#        """
from local_packages.nested_integer import NestedInteger


class NestedIterator:
    def __init__(self, nestedList: [NestedInteger]):
        def generator(nested: NestedInteger):
            if nested.isInteger():
                yield nested.getInteger()
            else:
                for i in nested.getList():
                    yield from generator(i)

        def listGenerator(nestedList: [NestedInteger]):
            for nested in nestedList:
                yield from generator(nested)

        self.generator = listGenerator(nestedList)
        self.val = None

    def next(self) -> int:
        return self.val

    def hasNext(self) -> bool:
        try:
            self.val = next(self.generator)
            return True
        except StopIteration:
            return False


# Your NestedIterator object will be instantiated and called as such:
# i, v = NestedIterator(nestedList), []
# while i.hasNext(): v.append(i.next())
for nestedList, expected in [
    ([[1, 1], 2, [1, 1]], [1, 1, 2, 1, 1]),
    ([1, [4, [6]]], [1, 4, 6]),
]:
    nintlist = NestedInteger(nestedList)
    i, actual = NestedIterator(nintlist.getList()), []
    while i.hasNext():
        actual.append(i.next())
    print("Flattened iterator output of", nestedList, "->", actual)
    assert actual == expected
